service/history/timerQueueProcessor.go

```diff
@@ -742,15 +745,31 @@ Update_History_Loop:
 		}
 
 		scheduleNewDecision := false
-		di, isRunning := msBuilder.GetPendingDecision(scheduleID)
-		if isRunning && msBuilder.isWorkflowExecutionRunning() {
-			// Add a decision task timeout event.
-			timeoutEvent := msBuilder.AddDecisionTaskTimedOutEvent(scheduleID, di.StartedID)
-			if timeoutEvent == nil {
-				// Unable to add DecisionTaskTimedout event to history
-				return &workflow.InternalServiceError{Message: "Unable to add DecisionTaskTimedout event to history."}
-			}
+		switch task.TimeoutType {
+		case int(workflow.TimeoutTypeStartToClose):
+			di, isRunning := msBuilder.GetPendingDecision(scheduleID)
+			if isRunning && msBuilder.isWorkflowExecutionRunning() {
+				// Add a decision task timeout event.
+				timeoutEvent := msBuilder.AddDecisionTaskTimedOutEvent(scheduleID, di.StartedID)
+				if timeoutEvent == nil {
+					// Unable to add DecisionTaskTimedout event to history
+					return &workflow.InternalServiceError{Message: "Unable to add DecisionTaskTimedout event to history."}
+				}
 
+				scheduleNewDecision = true
+			}
+		case int(workflow.TimeoutTypeScheduleToStart):
```

Comment:
Can we have a metric on each of these timeouts?

Reply:
added t.metricsClient.IncCounter(metrics.TimerTaskDecisionTimeoutScope, metrics.ScheduleToStartTimeoutCounter) and t.metricsClient.IncCounter(metrics.TimerTaskDecisionTimeoutScope, metrics.StartToCloseTimeoutCounter)